### PR TITLE
[XPU][Fix] Broadcast rank-0 token ids in TP sync instead of all_reduce(MIN)

### LINKED ISSUE #31011: [Bug] Intel XPU: TP token-id sync corrupted — xccl all_reduce(MIN/MAX) silently does SUM, breaking grammar/structured decoding.
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions.
- [x] Please use English.

### Describe the bug

On Intel XPU, `torch.distributed.all_reduce` with `ReduceOp.MIN` (or `MAX`) on integer tensors silently computes **SUM** instead — the xccl/oneCCL backend does not implement integer min/max reductions and falls back to SUM. (`SUM` itself is correct, which points to a backend op-support gap rather than a data issue.)

This corrupts SGLang's TP token-id synchronization. `Sampler._sync_token_ids_across_tp` in `python/sglang/srt/layers/sampler.py` syncs sampled token ids across TP ranks with `all_reduce(op=ReduceOp.MIN)`. It runs when grammar / regex / JSON constrained decoding is enabled, or when `SYNC_TOKEN_IDS_ACROSS_TP=1`.

**Impact:** In the normal case every TP rank samples the same token id, so a correct `MIN` is a no-op. On XPU the broken reduction instead returns `token_id * tp_size`, which typically lands outside the vocabulary. The TP ranks then desynchronize, and inference either produces garbage tokens or hangs/crashes.

**Scope:**
- Intel XPU (xccl/oneCCL) only. NVIDIA (nccl) computes MIN/MAX correctly and is unaffected.
- Triggered with `tp_size > 1` **and** grammar/regex/JSON constrained decoding, or `SYNC_TOKEN_IDS_ACROSS_TP=1`.
- Root cause is in the oneCCL backend (SUM works, MIN/MAX do not).

### Reproduction

**Minimal reproducer** — `torchrun --nproc-per-node=2 repro_xpu_sampler_token_id_sync.py`:
```
python
# repro_xpu_sampler_token_id_sync.py
import torch, torch.distributed as dist

dist.init_process_group(backend="xccl")
rank = dist.get_rank()
torch.xpu.set_device(rank)

# Every rank samples the SAME token id -> MIN must be a no-op.
token_ids = torch.tensor([42], dtype=torch.long, device=f"xpu:{rank}")
dist.all_reduce(token_ids, op=dist.ReduceOp.MIN)

# Correct backends print 42 on every rank.
# XPU/xccl prints 84 (== 42 * world_size), i.e. it computed SUM.
print(f"[rank {rank}] after all_reduce(MIN) = {token_ids.item()} (expected 42)")
```

[repro_xpu_sampler_token_id_sync.py](https://github.com/user-attachments/files/30213310/repro_xpu_sampler_token_id_sync.py)

### Environment

- SGLang: 0.5.16.dev (commit 4cec9ef, main)
- PyTorch: 2.12.0+xpu
- Distributed backend: xccl (oneCCL)
- Hardware: Intel Arc Pro B60 Graphics (8 devices), Intel Xeon 6980P
- OS: Ubuntu 25.04

### PR DESCRIPTION
**Pull request by dayanandav**
…e(MIN)

Intel XPU's xccl backend silently computes SUM for all_reduce with ReduceOp.MIN/MAX on integer tensors. This corrupts the TP token-id sync in Sampler._sync_token_ids_across_tp: when every rank samples the same id, the no-op MIN instead returns id * tp_size, which lands out of vocabulary and desyncs the TP ranks -> garbage output or a hang. Grammar/regex/JSON constrained decoding (and SYNC_TOKEN_IDS_ACROSS_TP=1) trigger this path.

On XPU, broadcast the group's first rank's token ids instead of all_reduce(MIN). This picks the same single deterministic winner while using only SUM-free collectives. All other platforms keep the original all_reduce(MIN) path.

Fixes #31011 .

## Motivation

On Intel XPU, `torch.distributed.all_reduce` with `ReduceOp.MIN` (or `MAX`) on integer tensors silently computes **SUM** — the xccl backend does not implement integer min/max and falls back to SUM.

This corrupts SGLang's TP token-id synchronization.
`Sampler._sync_token_ids_across_tp` (`python/sglang/srt/layers/sampler.py`) syncs sampled token ids across TP ranks with `all_reduce(op=ReduceOp.MIN)`, which runs when grammar/regex/JSON constrained decoding is enabled or when `SYNC_TOKEN_IDS_ACROSS_TP=1`. In the normal case every rank samples the same id, so `MIN` should be a no-op; on XPU the broken reduction instead returns `token_id * tp_size`, which typically exceeds vocabulary bounds. The TP ranks then desynchronize, producing garbage output or hanging.

## Modifications

Changes are isolated to `python/sglang/srt/layers/sampler.py`:
- Added an XPU detection flag (`self.token_id_sync_via_broadcast = is_xpu()`) in `Sampler.__init__`, matching the existing platform-detection pattern, and precomputed the source's **global** rank (`dist.get_global_rank(self.tp_sync_group, 0)`) since `dist.broadcast` expects a global rank.
- For XPU: replace `all_reduce(op=MIN)` with `dist.broadcast` of the group's first rank's token ids. This selects the same single deterministic winner `MIN` would, using only SUM-free collectives.
- For all other platforms: the original `all_reduce(op=ReduceOp.MIN)` path is unchanged.
- No API changes, no new environment variables, no configuration additions.

## Accuracy Tests

- Unit reproducer, TP=2 on XPU, identical ids `[42, 42]`:
  - Before: sync produced `[84, 84]` (out-of-vocab) → desync.
  - After: sync produces `[42, 42]`.
- Grammar/JSON-constrained GSM8K on <Qwen/Qwen3-VL-2B-Thinking> (XPU TP=2):
  - Before: score collapsed to ~0 from invalid tokens.
  - After: matches the CUDA TP=2 reference within noise.
- CUDA regression with token sync enabled: unchanged (non-XPU path untouched).

## Speed Tests and Profiling

The sync path only activates when `tp_sync_group` is set (grammar constraints or the opt-in env var), not on the default hot path. `broadcast` of a `[bs]` int64 tensor is equivalent in cost to the `all_reduce` it replaces; end-to-end TTFT/TPOT impact is within noise (< 0.5%). <microbenchmark μs before vs. after, XPU TP=2>.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #30452309076](https://github.com/sgl-project/sglang/actions/runs/30452309076)
Latest PR Test (Extra): :x: [Run #30452308734](https://github.com/sgl-project/sglang/actions/runs/30452308734)

### CODE REVIEW

**Comment by dayanandav** [comment]
@siju-samuel requested for review.

**Review by siju-samuel — CHANGES_REQUESTED** [review]
i think we dont need this for xpu, we already support min/max 
https://github.com/intel/torch-xpu-ops/blob/main/test/xpu/distributed/test_c10d_ops_xccl.py#L213

**Comment by dayanandav** [comment]
Issue is resolved with latest oneccl version, so can be closed this PR.
